### ISSUE TITLE
Error messages for the same problem are different over http and amqpws

### ISSUE THREAD

**Issue by maestroblue** [Azure/azure-iot-sdk-node#777]
**package.json:**
"azure-iot-common": "1.11.2"
"azure-iot-device": "1.12.2"
"azure-iot-device-amqp": "1.11.2"
"azure-iot-device-http": "1.11.2"

Problem:
We use amqpWs in our production to communicate with IoT. When a device is deleted in IoT and the sdk tries to open a connection and send a message, we get the following error with amqpWs:

> { [UnauthorizedError: Device <DEVICE_ID> at IotHub blue-ci-iot-hub not registered.] name: 'UnauthorizedError' }

This error does not indicate what the actual problem is, unless we parse the error string and look for 'not registered'. Our workaround the problem is to use http sdk to check for device registration, if we got an error once we opened the client (client.open()) and tried to send a message (client.sendEvent()). When we use the http protocol, the error returned by the sdk is as follows:

> { [DeviceNotFoundError: Could not send message: Not Found] name: 'DeviceNotFoundError' }

This leads to two problems for us, albeit not being very common is likely to be soon as we are enrolling more customers to use this solution.
1) The azure-iot-device-http sdk uses port 5671, which is likely to be blocked in most of the customer firewall configurations. In this case, our current workaround fails as well.
2) There are two different error messages for the same issue, which leads us to check for the actual problem through this indirect way. 

As a side note, we are trying to determine if a device is disabled or deleted from IoT (a recent requirement), and we are not sure how we can check for that state and differentiate between deleted and disabled. Any suggestion on this one is much appreciated.

Thank you for your time!

AB#7366646

**Comment by anthonyvercolano** [comment]
I'll comment on your three questions.

1) I'm pretty sure that you mean that azure-iot-device-***AMQP*** uses port 5671.  HTTPS uses 443 typically.  Can you double check on this?  Even so, I'm not really sure what the question is here.

2). We are not likely as a rule to give you much information on the device.  In general we would be vague because we don't want to help an antagonist sniffing around your hub.  If anything, I would say that we are giving too much information on the responses.  But, why is looking for the 'not registered' string bad?  I will check with the service folk how they feel about the different responses but at this point I doubt they would change the behavior of either gateway (amqp/http) as this could be thought of as a breaking change.  I'll post an answer when I get it.

Further, I'm not really sure why it matters.  Your device app receives either one of these errors.  There really isn't anything that it can do at this point.  Any of these errors is fairly non-recoverable at this point.  You could try to roll the key I guess?  The device app is not likely to want to create a new device, correct?  Seems at though it has to wait for some reconfiguration after providing some sort of notification that will be noticed by an administrator?  However, without a device you're going to have to roll your own notification mechanism?

3) Disabled/Enabled.  Well you could find this out by using the service client.  However, I'm thinking you don't want to have your devices knowing the private key for the entire hub.  I'll ask around but I don't think that this is possible.  If I get something I'll post it here.

**Comment by gulipek95** [comment]
Thank you for the fast response! 

For the first point, more than a question my intention was to point out the problem we have with the workaround, though I will have to doublecheck as you point out. 

That's a good point about error details, we also try to avoid giving information on the API's we have for our sdk, but I guess when you are on the consumer end it becomes more pressing when you want to automate a certain workflow. In our case we need to differentiate between not registered vs any other error, since for the former, (this is the part we automate) if we receive that then we delete the connectionString (and any other connection information) from the database. This puts the edge device in a clean/unclaimed state.

In return the thick client indicates that the device's subscription has been removed, and the customer will need to create either a new account or that was the intention and we're done with the workflow. To give a bit more detail, we have a web client that device could be deleted from, and if that's the case we propagate that to the edge device which is a deployment on a private/ closed network usually, and there is a desktop client that displays the state in near-real time.

If however the error is not 'device not registered', then we do not nuke the state, but we retry the connection - for the error is likely a recoverable issue at some point. We still close the client & remove the references. I hope I explained the workflow a bit better.

As for not parsing a string, I guess the reasoning was that a string is more likely to change than the error code so we considered it as more error-prone. The other reason is that (not likely but) the string might exist in another error, whereas the error codes are specific to the error, at least the kind that makes us nuke the state :) 

For the disabled/enabled state, if we are in a disabled state, we won't nuke the information but we won't keep retrying either. I am guessing we'll need some management on the thick client that requires the customer to execute some action once the device is enabled, but the main part is knowing that we are in that state. Any information on this would be much appreciated, if it is not doable I'll inform the team accordingly.

Thank you for your time Anthony!
